In [26]:
import numpy as np

In [92]:
import math
import random

In [36]:
"""
Shreds each document and returns a list with the counts for each character
Index 0 is characters 'a' count, Index 1 is characters 'b' count, etc:
Index 26 is the count for an empty space
"""
def shred_data(doc_list):
    p = [0] * 27
    val = 97
    
    for file in doc_list:
        with open (file,encoding='utf-8') as f:
        # TODO: add your code here
            for line in f:
                for c in line:
                    ascii_c = ord(c)  
                    # for every character check if in range and increase count
                    if ascii_c >= ord('a') and ascii_c <= ord('z'):
                        p[ascii_c - val] += 1
                    elif chr(ascii_c) == ' ':
                        p[26] += 1
                          

    return p
    

In [139]:
"""
Calculate prior probabilities for each label

"""
def calc_prior(N, train_count, smoothing):
    priors = []
    labels = ['e','s','j']
    i = 0
    for label in labels:
        n = (train_count[i] + smoothing)/(N + 3 * smoothing)
        
        s = 'P(y = ' + label + ') = ' + str(n)
        print(s)
        priors.append(n)
    
    return priors

In [39]:
"""
Calculate the class conditional probabilities for each character
"""
def calc_likelihood(lang_vector, smoothing, N):
    theta = [0] * 27
    
    for i in range(27):
        theta[i] = (lang_vector[i] + 0.5)/(np.array(lang_vector).sum() + 0.5 * 27)
    return theta

In [76]:
"""
Calculate the likelihood of p(x|y)
"""
def calc_test_l(theta, bow):
    x = 0
    for i in range(27):
        x = x + bow[i] * math.log(theta[i])
    return x


# 2.1)
 

In [140]:
train_count = [10,10,10] # list that includes training counts for e, s, j in that order
priors = calc_prior(30, train_count , 0.5)

P(y = e) = 0.3333333333333333
P(y = s) = 0.3333333333333333
P(y = j) = 0.3333333333333333


# 2.2)

In [141]:
labels = ['e', 's', 'j']
lang_dict = {} # stores bag of words for each label for all training data

for label in labels:
    file_list = []
    for i in range(10):
        file = "languageID" + "/" + label + str(i) + ".txt"
        file_list.append(file)
    lang_dict[label] = shred_data(file_list)
    
N = 0
for label in lang_dict:
    N += np.array(lang_dict[label]).sum()
    
# get english conditional probabilty
theta_e = calc_likelihood(lang_dict['e'], 0.5, N)
print("Class conditional probability for english")
print(theta_e)

Class conditional probability for english
[0.0601685114819098, 0.011134974392863043, 0.021509995043779945, 0.021972575582355856, 0.1053692383941847, 0.018932760614571286, 0.017478936064761277, 0.047216256401784236, 0.055410540227986124, 0.001420783082768875, 0.0037336857756484387, 0.028977366595076822, 0.020518751032545846, 0.057921691723112505, 0.06446390219725756, 0.01675202378985627, 0.0005617049396993227, 0.053824549810011564, 0.06618205848339666, 0.08012555757475633, 0.026664463902197257, 0.009284652238559392, 0.015496448042293078, 0.001156451346439782, 0.013844374690236246, 0.0006277878737815959, 0.1792499586981662]


# 2.3)

In [142]:
print("Class conditional probability for japanese")
theta_j = calc_likelihood(lang_dict['j'], 0.5, N)
print(theta_j)
theta_s = calc_likelihood(lang_dict['s'], 0.5, N)
print("Class conditional probability for spanish")
print(theta_s)

Class conditional probability for japanese
[0.1317656102589189, 0.010866906600510151, 0.005485866033054963, 0.01722631818022992, 0.06020475907613823, 0.003878542227191726, 0.014011670568503443, 0.03176211607673224, 0.09703343932352633, 0.0023411020650616725, 0.05740941332681086, 0.001432614696530277, 0.03979873510604843, 0.05671057688947902, 0.09116321324993885, 0.0008735455466648031, 0.00010482546559977637, 0.04280373178657535, 0.0421747789929767, 0.056990111464411755, 0.07061742199238269, 0.0002445927530661449, 0.01974212935462455, 3.4941821866592126e-05, 0.01415143785596981, 0.00772214263251686, 0.12344945665466997]
Class conditional probability for spanish
[0.10456045141993771, 0.008232863618143134, 0.03752582405722919, 0.039745922111559924, 0.1138108599796491, 0.00860287996053159, 0.0071844839813758445, 0.0045327001942585795, 0.049859702136844375, 0.006629459467793161, 0.0002775122567913416, 0.052943171656748174, 0.02580863988159477, 0.054176559464709693, 0.07249236841293824, 0.02

# 2.4)

In [131]:
file = "languageID/e10.txt"
bow = shred_data([file])
print(bow)

[164, 32, 53, 57, 311, 55, 51, 140, 140, 3, 6, 85, 64, 139, 182, 53, 3, 141, 186, 225, 65, 31, 47, 4, 38, 2, 498]


# 2.5)

In [133]:
# probabilities below are stored in the log space

print("P(x|y = e)")
e_likelihood = calc_test_l(theta_e, bow)
print(e_likelihood)
print("P(x|y = s)")
s_likelihood = calc_test_l(theta_s, bow)
print(s_likelihood)
print("P(x|y = j)")
j_likelihood = calc_test_l(theta_j, bow)
print(j_likelihood)

P(x|y = e)
-7841.865447060635
P(x|y = s)
-8467.282044010557
P(x|y = j)
-8771.433079075032


# 2.6

Since p(y = e|x) = -7842 is the largest value or probability in the log space, the predicted class label for this document is english which is correct

In [134]:
# P(y = e|x) = P(x|y=e) * P(e)
# Since arithmetic is done in log space we add log(P(e)) to log(P(x|y=e))

print("P(y = e|x)")
print(e_likelihood + np.log(priors[0]))
print("P(y = s|x)")
print(s_likelihood + np.log(priors[1]))
print("P(y = j|x)")
print(j_likelihood + np.log(priors[2]))

# English is the predicted class label

P(y = e|x)
-7842.964059349303
P(y = s|x)
-8468.380656299225
P(y = j|x)
-8772.5316913637


# 2.7

See confusion matrix in HW4 document. The Naive Bayes Classifier had 100% accuracy with no False labels.


In [135]:
#Evaluation
TE = 0
FSE = 0
FJE = 0
TS = 0
FES = 0
FJS = 0
TJ = 0
FEJ = 0
FSJ = 0

label = "e"

for i in range(10):
    file = "languageID" + "/" + label + str(i+10) + ".txt"
    bow = shred_data([file])
    e_likelihood = calc_test_l(theta_e, bow)
    s_likelihood = calc_test_l(theta_s, bow)
    j_likelihood = calc_test_l(theta_j, bow)
     
    
    e_prob = e_likelihood + np.log(priors[0])
    s_prob = s_likelihood + np.log(priors[1])
    j_prob = j_likelihood + np.log(priors[2])

    if(e_prob > s_prob and e_prob > j_prob):
        TE += 1
    elif(s_prob > e_prob and s_rpob > j_prob):
        FSE += 1
    else:
        FJE += 1

label = "s"

for i in range(10):
    file = "languageID" + "/" + label + str(i+10) + ".txt"
    bow = shred_data([file])
    e_likelihood = calc_test_l(theta_e, bow)
    s_likelihood = calc_test_l(theta_s, bow)
    j_likelihood = calc_test_l(theta_j, bow)
        
    e_prob = e_likelihood + np.log(priors[0])
    s_prob = s_likelihood + np.log(priors[1])
    j_prob = j_likelihood + np.log(priors[2])
    

    if(s_prob > e_prob and s_prob > j_prob):
        TS += 1
    elif(e_prob > s_prob and e_rpob > j_prob):
        FES += 1
    else:
        FJS += 1

label = "j"

for i in range(10):
    file = "languageID" + "/" + label + str(i+10) + ".txt"
    bow = shred_data([file])
    e_likelihood = calc_test_l(theta_e, bow)
    s_likelihood = calc_test_l(theta_s, bow)
    j_likelihood = calc_test_l(theta_j, bow)
        
    e_prob = e_likelihood + np.log(priors[0])
    s_prob = s_likelihood + np.log(priors[1])
    j_prob = j_likelihood + np.log(priors[2])
    
   
    if(j_prob > e_prob and j_prob > s_prob):
        TJ += 1
    elif(e_prob > s_prob and e_rpob > j_prob):
        FEJ += 1
    else:
        FSJ += 1
        
print("True English: ", str(TE))
print("True Spanish: ", str(TS))
print("True Japanese: ", str(TJ))


True English:  10
True Spanish:  10
True Japanese:  10


# 2.8

In [136]:
"""
Shuffle every line and every character in every line of document
See shuffle.txt for results
"""
def shuffle(doc):
    with open (doc,encoding='utf-8') as f:
        lines = f.readlines() 
        
    random.shuffle(lines)
    new_lines = []
    for line in lines:
        line = "".join(random.sample(line, len(line)))
        new_lines.append(line)
    with open('shuffle.txt', 'w') as f: 
        f.writelines(new_lines) 
                
                

In [137]:
#Shuffling
test_doc = "languageID/e10.txt"
shuffle(test_doc)

In [138]:
bow = shred_data(["shuffle.txt"])
print("P(x|y = e)")
e_likelihood = calc_test_l(theta_e, bow)
print(e_likelihood)
print("P(x|y = s)")
s_likelihood = calc_test_l(theta_s, bow)
print(s_likelihood)
print("P(x|y = j)")
j_likelihood = calc_test_l(theta_j, bow)
print(j_likelihood)

print("English")
print(e_likelihood + np.log(priors[0]))
print("Spanish")
print(s_likelihood + np.log(priors[1]))
print("Japanese")
print(j_likelihood + np.log(priors[2]))

P(x|y = e)
-7841.865447060635
P(x|y = s)
-8467.282044010557
P(x|y = j)
-8771.433079075032
English
-7842.964059349303
Spanish
-8468.380656299225
Japanese
-8772.5316913637
